pkg/controller/controlplane/reconciler.go

```diff
@@ -0,0 +1,195 @@
+)
+
+const (
+	// EventControlPlaneReconciliation an event reason to describe control plane reconciliation.
+	EventControlPlaneReconciliation string = "ControlPlaneReconciliation"
+	// EventControlPlaneDeletion an event reason to describe control plane deletion.
+	EventControlPlaneDeletion string = "ControlPlaneDeletion"
+)
+
+type reconciler struct {
+	logger   logr.Logger
+	actuator Actuator
+
+	ctx      context.Context
+	client   client.Client
+	recorder record.EventRecorder
+}
+
+// NewReconciler creates a new reconcile.Reconciler that reconciles
+// controlplane resources of Gardener's `extensions.gardener.cloud` API group.
+func NewReconciler(mgr manager.Manager, actuator Actuator) reconcile.Reconciler {
+	return &reconciler{
+		logger:   log.Log.WithName(ControllerName),
+		actuator: actuator,
+		recorder: mgr.GetRecorder(ControllerName),
+	}
+}
+
+func (r *reconciler) InjectFunc(f inject.Func) error {
+	return f(r.actuator)
+}
+
+func (r *reconciler) InjectClient(client client.Client) error {
+	r.client = client
+	return nil
+}
+
+func (r *reconciler) InjectStopChannel(stopCh <-chan struct{}) error {
+	r.ctx = util.ContextFromStopChannel(stopCh)
+	return nil
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	cp := &extensionsv1alpha1.ControlPlane{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, cp); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster, err := extensionscontroller.GetCluster(r.ctx, r.client, cp.Namespace)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if cp.DeletionTimestamp != nil {
+		return r.delete(r.ctx, cp, cluster)
+	}
+	return r.reconcile(r.ctx, cp, cluster)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, cp *extensionsv1alpha1.ControlPlane, cluster *extensionscontroller.Cluster) (reconcile.Result, error) {
+	if err := extensionscontroller.EnsureFinalizer(ctx, r.client, FinalizerName, cp); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	operationType := gardencorev1alpha1helper.ComputeOperationType(cp.ObjectMeta, cp.Status.LastOperation)
+	if err := r.updateStatusProcessing(ctx, cp, operationType, "Reconciling the controlplane"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the reconciliation of controlplane", "controlplane", cp.Name)
+	r.recorder.Event(cp, corev1.EventTypeNormal, EventControlPlaneReconciliation, "Reconciling the controlplane")
+	if err := r.actuator.Reconcile(ctx, cp, cluster); err != nil {
+		msg := "Error reconciling controlplane"
+		r.updateStatusError(ctx, extensionscontroller.ReconcileErrCauseOrErr(err), cp, operationType, msg)
+		r.logger.Error(err, msg, "controlplane", cp.Name)
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	if kutil.HasMetaDataAnnotation(&cp.ObjectMeta, gardencorev1alpha1.GardenerOperation, gardencorev1alpha1.GardenerOperationReconcile) {
```

Comment:
Can you remove this? It's not needed for the ControlPlane controller.

Reply:
Sure, done